### PR TITLE
Learn rate schedulers

### PR DESCRIPTION
**Pull request by bluesdog164**
This is my first open source project merge request. Sorry if it is rough around the edges:
- Added support to Pytorch learning rate schedulers as callbacks;
- Added WarmRestartLR from "SGDR: Stochastic Gradient Descent with Warm Restarts" https://arxiv.org/abs/1608.03983  (Pytorch issue  #3740);
- Added Dockerfile and docker-compose files to allow testing skorch without installing;

### CODE REVIEW

File: skorch/lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+class WarmRestartLR(_LRScheduler):
+    """Sets the learning rate of each parameter group according to
+    stochastic gradient descent with warm restarts (SGDR) policy. The
+    policy simulates periodic warm restarts of SGD, where in each restart the
+    learning rate is initialize to some value and is scheduled to decrease,
+    as described in the paper
+    `"Stochastic Gradient Descent with Warm Restarts"
+    <https://arxiv.org/pdf/1608.03983.pdf>`_
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
```
Comment (ottonemo) [inline skorch/lr_scheduler.py]: "Wrapped optimizer"? The input is just a plain optimizer, right?

File: skorch/lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+class WarmRestartLR(_LRScheduler):
+    """Sets the learning rate of each parameter group according to
+    stochastic gradient descent with warm restarts (SGDR) policy. The
+    policy simulates periodic warm restarts of SGD, where in each restart the
+    learning rate is initialize to some value and is scheduled to decrease,
+    as described in the paper
+    `"Stochastic Gradient Descent with Warm Restarts"
+    <https://arxiv.org/pdf/1608.03983.pdf>`_
+    Args:
```
Comment (ottonemo) [inline skorch/lr_scheduler.py]: Can you format this in the numpy doc style so that it gets rendered correctly in the online documentation? (See [section 4](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections)).

File: skorch/tests/test_lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        self.dropout = nn.Dropout(0.5)
+        self.dense1 = nn.Linear(num_units, 10)
+        self.output = nn.Linear(10, 2)
+
+    def forward(self, X):
+        X = self.nonlin(self.dense0(X))
+        X = self.dropout(X)
+        X = self.nonlin(self.dense1(X))
+        X = F.softmax(self.output(X), dim=-1)
+        return X
```
Comment (ottonemo) [inline skorch/tests/test_lr_scheduler.py]: This whole class can be removed. You can simply use `classifier_module` as a fixture in your tests which is equivalent to `SchedulerTestNet`. For example `def test_lr_callback_init_policies(self):` would change to ` def test_lr_callback_init_policies(self, classifier_module):` and the invocation of `SchedulerTestNet` in subsequent tests would be substituted with `classifier_module`. This way you don't need to copy code.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ADD requirements-dev.txt requirements-dev.txt
+RUN pip3 install -r requirements-dev.txt
+
+ADD . /app
+RUN pip3 install -e .
+
+ADD http://www.manythings.org/anki/fra-eng.zip /app/examples/translation/data/
+
+RUN unzip /app/examples/translation/data/fra-eng.zip -d /app/examples/translation/data/
+RUN mv /app/examples/translation/data/fra.txt /app/examples/translation/data/eng-fra.txt
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply Dockerfile]: First, thank you for your thorough review, and sorry about the late response.

About this line, this is just to make the https://github.com/dnouri/skorch/blob/master/examples/translation/translation.ipynb ready to be executed inside the jupyter notebook.

Do you thing that we should remove that?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+EXPOSE 8888
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply Dockerfile]: Done

File: skorch/__init__.py [hunk 6]
```diff
@@ -16,6 +17,7 @@
     'NeuralNetClassifier',
     'NeuralNetRegressor',
     'callbacks',
+    'lr_scheduler'
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply skorch/__init__.py]: Done. Removed this line

File: skorch/lr_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
+import numpy as np
+from torch.optim.lr_scheduler import *
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply skorch/lr_scheduler.py]: Done

File: skorch/lr_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+        if self.policy == 'warm_restart':
+            min_lr = kwargs.get('min_lr', 1e-6)
+            max_lr = kwargs.get('max_lr', 0.05)
+            base_period = kwargs.get('base_period', 10)
+            period_mult = kwargs.get('period_mult', 2)
+            return WarmRestartLR(
+                optimizer, min_lr, max_lr, base_period,
+                period_mult, last_epoch
+            )
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply skorch/lr_scheduler.py]: True. That way is a lot better. I did have to change some other parts, but I think it is done.

File: skorch/tests/test_lr_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        assert any(list(map(
+            lambda x: isinstance(
+                getattr(x[1], '_lr_scheduler', None), instance),
+            net.callbacks_
+        )))
+
+class TestWarmRestartLR():
+    def setup_class(self):
+        self.net = SchedulerTestNet()
+        self.opt = SGD(self.net.parameters(), lr = 0.05)
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply skorch/tests/test_lr_scheduler.py]: Done.

File: skorch/tests/test_lr_scheduler.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        scheduler = WarmRestartLR(
+            self.opt, min_lr, max_lr, base_period, period_mult
+        )
+        self._test(scheduler, targets, epochs)
+
+    def _test(self, scheduler, targets, epochs=10):
+        for epoch in range(epochs):
+            scheduler.step(epoch)
+            for param_group, target in zip(self.opt.param_groups, targets):
+                 pytest.approx(target[epoch], param_group['lr'])
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply skorch/tests/test_lr_scheduler.py]: Changed that and founded some bugs. Thanks.

File: skorch/lr_scheduler.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    stochastic gradient descent with warm restarts (SGDR) policy. The
+    policy simulates periodic warm restarts of SGD, where in each restart the
+    learning rate is initialize to some value and is scheduled to decrease,
+    as described in the paper
+    `"Stochastic Gradient Descent with Warm Restarts"
+    <https://arxiv.org/pdf/1608.03983.pdf>`_
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
+        min_lr (float or list): minimum learning rate for each param groups.
+        max_lr (float or list): maximum learning rate for each param groups.
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply skorch/lr_scheduler.py]: Missed that. Gonna change that later.

File: skorch/tests/test_lr_scheduler.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+from torch.optim import SGD
+import torch
+
+from skorch.lr_scheduler import WarmRestartLR, LRScheduler
+from sklearn.datasets import make_classification
+from skorch.net import NeuralNetClassifier
+import numpy as np
+import pytest
+
+#TODO: Copying from pytorch test/test_optim. Should be working
```
[the remark this answers is not in the document]
Comment (bluesdog164) [reply skorch/tests/test_lr_scheduler.py]: Sorry. Missed that comment after solving it. Removed from the file.

File: skorch/lr_scheduler.py (hunk 1 above)
Comment (bluesdog164) [reply skorch/lr_scheduler.py]: That's the way how Pytorch docs referred to this argument (http://pytorch.org/docs/master/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau). Changed that

File: skorch/lr_scheduler.py (hunk 2 above)
Comment (bluesdog164) [reply skorch/lr_scheduler.py]: Sure.

File: skorch/tests/test_lr_scheduler.py (hunk 3 above)
Comment (bluesdog164) [reply skorch/tests/test_lr_scheduler.py]: Done.

**Comment by bluesdog164** [comment]
Again, thank you for your thorough review, and sorry about the late response.

I'm still editing this PR after the requested changes, so I added a WIP in the title.
Also, since I did find some bugs and coverage now is below 90%, I'm adding some new test.

**Comment by bluesdog164** [comment]
I think it's done @ottonemo. Could you please review it again? Thanks!

**Comment by ottonemo** [comment]
Thanks for addressing the changes. I reviewed your revised code and it seems good to merge. We'll move the `lr_scheduler` in a separate callback submodule.

Regarding the `Dockerfile`: We'll remove the example code. The idea would be to have one generic Dockerfile that provides a working container with skorch and then one could write simple Dockerfiles that extend on that container and provide example specific instructions.

Thanks for your contribution!
